<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>chavi: Consensus Haplotype Detection and Visualization • chavi</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><meta property="og:title" content="chavi: Consensus Haplotype Detection and Visualization">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">chavi</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/chavi.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>chavi: Consensus Haplotype Detection and Visualization</h1>
                        <h4 class="author">Anna Plantinga</h4>
            
            <h4 class="date">2018-03-10</h4>
          </div>

    
    
<div class="contents">
<p>chavi is a package that identifies and visualizes long shared IBS between case haplotypes. It is intended to be used in the setting where a particular genomic site is of interest (e.g., a mutation site) and the region around that site is expected to be highly conserved among carriers. We also assume that only one haplotype per case subject carries the feature.</p>
<p>Vocabulary used throughout this document:</p>
<ul>
<li>“Case subjects” are those that carry the site of interest on one chromosome copy. In the case where the feature of interest is a mutation known to be causal for a disease, case subjects are those with the disease.</li>
<li>“Consensus risk haplotype” refers to the stretch of genome shared identically between (nearly) all cases around the site of interest.</li>
<li>“Risk haplotype” refers to the copy of the chromosome for each subject that carries the feature of interest.</li>
</ul>
<div id="step-0-install-chavi" class="section level2">
<h2 class="hasAnchor">
<a href="#step-0-install-chavi" class="anchor"></a>Step 0: Install chavi</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(devtools)
<span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"aplantin/chavi"</span>)
<span class="co">#&gt; Downloading GitHub repo aplantin/chavi@master</span>
<span class="co">#&gt; from URL https://api.github.com/repos/aplantin/chavi/zipball/master</span>
<span class="co">#&gt; Installing chavi</span>
<span class="co">#&gt; '/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file  \</span>
<span class="co">#&gt;   --no-environ --no-save --no-restore --quiet CMD INSTALL  \</span>
<span class="co">#&gt;   '/private/var/folders/yw/fmxy3ljs5yxf1frzv8b2fbhh0000gn/T/RtmpdFLzaZ/devtools3782667d2e1e/aplantin-chavi-7b31b51'  \</span>
<span class="co">#&gt;   --library='/Library/Frameworks/R.framework/Versions/3.4/Resources/library'  \</span>
<span class="co">#&gt;   --install-tests</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Reloading installed chavi</span></code></pre></div>
</div>
<div id="step-1-load-and-prepare-data" class="section level2">
<h2 class="hasAnchor">
<a href="#step-1-load-and-prepare-data" class="anchor"></a>Step 1: Load and prepare data</h2>
<p>We assume that the available data include a phased VCF file and a text file with one case subject identifier per line. The VCF file may include other subjects in addition to cases; these will be filtered out for the remainder of the analysis.</p>
<p>The VCF file is converted to a “haplotype” file. Rows are genetic markers, and there is a column labeled POS identifying the position of the genetic markers (in bp or cM). All of the markers must be on a single chromosome. The remaining columns contain the data for each case haplotype. The haplotype identifiers are SUBJ.1 and SUBJ.2 for the first and second copy of the chromosome for the case subject with identifier SUBJ. Other columns may be included and will be ignored.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">vcf.file &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"sim-haps.vcf"</span>, <span class="dt">package =</span> <span class="st">"chavi"</span>)
case.id.file &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"sim-case-ids.txt"</span>, <span class="dt">package =</span> <span class="st">"chavi"</span>)
haps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/vcf2hap.html">vcf2hap</a></span>(vcf.file, case.id.file)</code></pre></div>
<p>The haplotype file looks like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">haps[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]
<span class="co">#&gt;   POS SUBJID1.1 SUBJID1.2 SUBJID2.1 SUBJID2.2</span>
<span class="co">#&gt; 1 0.1         0         1         1         0</span>
<span class="co">#&gt; 2 0.2         1         1         0         1</span>
<span class="co">#&gt; 3 0.3         0         0         0         1</span>
<span class="co">#&gt; 4 0.4         1         1         1         1</span>
<span class="co">#&gt; 5 0.5         1         1         1         1</span></code></pre></div>
<p>After the haplotype file is loaded, we need to identify which copy of the chromosome carries the feature of interest for each case subject. This procedure assumes there is long IBS shared among case subjects at that site. The data were generated such that the site of interest is from 2.43 to 2.47 cM.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ibs.file &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"sim-ibs.txt"</span>, <span class="dt">package =</span> <span class="st">"chavi"</span>)
riskhaps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/id_riskhap.html">id_riskhap</a></span>(ibs.file, case.id.file, <span class="dt">site.start =</span> <span class="fl">2.43</span>, <span class="dt">site.end =</span> <span class="fl">2.47</span>)
riskhaps
<span class="co">#&gt;  [1] "SUBJID1.2"  "SUBJID2.2"  "SUBJID3.2"  "SUBJID4.1"  "SUBJID5.1" </span>
<span class="co">#&gt;  [6] "SUBJID6.2"  "SUBJID7.1"  "SUBJID8.1"  "SUBJID9.1"  "SUBJID10.1"</span></code></pre></div>
<p>This function tells us that Haplotype 2 of Subjects 1, 2, 3, and 6 carry the feature of interest, whereas for the remaining subjects, Haplotype 1 carries the feature.</p>
</div>
<div id="step-2-identify-consensus-risk-haplotype" class="section level2">
<h2 class="hasAnchor">
<a href="#step-2-identify-consensus-risk-haplotype" class="anchor"></a>Step 2: Identify consensus risk haplotype</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cons.out &lt;-<span class="st"> </span><span class="kw"><a href="../reference/get_consensus.html">get_consensus</a></span>(haps, riskhaps, <span class="dt">site.start =</span> <span class="fl">2.43</span>, <span class="dt">site.end =</span> <span class="fl">2.47</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>The consensus haplotype file contains […].</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(cons.out<span class="op">$</span>consensus)
<span class="co">#&gt;          id start end hapID</span>
<span class="co">#&gt; 1 SUBJID1.2   1.8 2.8     1</span>
<span class="co">#&gt; 2 SUBJID2.2   1.8 2.8     1</span>
<span class="co">#&gt; 3 SUBJID3.2   1.8 2.8     1</span>
<span class="co">#&gt; 4 SUBJID4.1   1.8 2.8     1</span>
<span class="co">#&gt; 5 SUBJID5.1   1.8 2.8     1</span>
<span class="co">#&gt; 6 SUBJID6.2   1.8 2.8     1</span></code></pre></div>
<p>We can summarize this data frame as the longest haplotype carried by each case subject:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/consensus_summary.html">consensus_summary</a></span>(cons.out<span class="op">$</span>consensus)
<span class="co">#&gt;            id start end</span>
<span class="co">#&gt; 1   SUBJID1.2   1.7 3.0</span>
<span class="co">#&gt; 2  SUBJID10.1   1.0 3.1</span>
<span class="co">#&gt; 3   SUBJID2.2   1.2 4.0</span>
<span class="co">#&gt; 4   SUBJID3.2   1.2 4.0</span>
<span class="co">#&gt; 5   SUBJID4.1   1.4 3.1</span>
<span class="co">#&gt; 6   SUBJID5.1   1.7 3.7</span>
<span class="co">#&gt; 7   SUBJID6.2   1.2 3.8</span>
<span class="co">#&gt; 8   SUBJID7.1   1.0 3.8</span>
<span class="co">#&gt; 9   SUBJID8.1   1.1 3.3</span>
<span class="co">#&gt; 10  SUBJID9.1   2.2 3.4</span></code></pre></div>
<p>The split locations are also stored in the output of get_consensus().</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(cons.out<span class="op">$</span>splits)
<span class="co">#&gt; [1] 1.3 1.8 2.8 2.9</span></code></pre></div>
</div>
<div id="step-3-visualize-consensus-risk-haplotype" class="section level2">
<h2 class="hasAnchor">
<a href="#step-3-visualize-consensus-risk-haplotype" class="anchor"></a>Step 3: Visualize consensus risk haplotype</h2>
<p>The simplest plot is the haplotypes alone.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_consensus.html">plot_consensus</a></span>(cons.out<span class="op">$</span>consensus, <span class="dt">symbols =</span> F, 
               <span class="dt">label.size =</span> <span class="fl">0.75</span>)</code></pre></div>
<p><img src="chavi_files/figure-html/plot-consensus-1-1.png" width="672"></p>
<p>For the sake of those who are colorblind (or even just clearer visualization for everyone else), symbols can be added to the plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_consensus.html">plot_consensus</a></span>(cons.out<span class="op">$</span>consensus, <span class="dt">symbols =</span> T, 
               <span class="dt">label.size =</span> <span class="fl">0.75</span>)</code></pre></div>
<p><img src="chavi_files/figure-html/plot-consensus-2-1.png" width="672"></p>
<p>If the genomic position of the feature of interest is specified, they will be drawn as solid vertical lines on the plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_consensus.html">plot_consensus</a></span>(cons.out<span class="op">$</span>consensus, <span class="dt">site.start =</span> <span class="fl">2.43</span>, <span class="dt">site.end =</span> <span class="fl">2.47</span>, <span class="dt">symbols =</span> T, 
               <span class="dt">label.size =</span> <span class="fl">0.75</span>)</code></pre></div>
<p><img src="chavi_files/figure-html/plot-consensus-3-1.png" width="672"></p>
<p>If the split locations are also included, they are drawn as dashed vertical lines. These demarcate the locations where a single group of subjects splits into two based on their allele at the split position.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_consensus.html">plot_consensus</a></span>(cons.out<span class="op">$</span>consensus, cons.out<span class="op">$</span>splits, 
               <span class="dt">site.start =</span> <span class="fl">2.43</span>, <span class="dt">site.end =</span> <span class="fl">2.47</span>, <span class="dt">symbols =</span> T, 
               <span class="dt">label.size =</span> <span class="fl">0.75</span>)</code></pre></div>
<p><img src="chavi_files/figure-html/plot-consensus-4-1.png" width="672"></p>
<p>If strong color preferences are present, colors may be specified explicitly.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_consensus.html">plot_consensus</a></span>(cons.out<span class="op">$</span>consensus, cons.out<span class="op">$</span>splits, 
               <span class="dt">site.start =</span> <span class="fl">2.43</span>, <span class="dt">site.end =</span> <span class="fl">2.47</span>, <span class="dt">symbols =</span> T, 
               <span class="dt">colors =</span> <span class="kw">c</span>(<span class="st">"red"</span>, <span class="st">"blue"</span>, <span class="st">"green"</span>), 
               <span class="dt">label.size =</span> <span class="fl">0.75</span>)</code></pre></div>
<p><img src="chavi_files/figure-html/plot-consensus-5-1.png" width="672"></p>
<p>If external grouping is important (e.g., some of the cases are from one population and some are from another), subjects may be ordered by group and group membership may be specified.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plot_consensus.html">plot_consensus</a></span>(cons.out<span class="op">$</span>consensus, cons.out<span class="op">$</span>splits, 
               <span class="dt">site.start =</span> <span class="fl">2.43</span>, <span class="dt">site.end =</span> <span class="fl">2.47</span>, <span class="dt">symbols =</span> T, 
               <span class="dt">groups =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>), 
               <span class="dt">group.names =</span> <span class="kw">c</span>(<span class="st">"Pop 1"</span>, <span class="st">"Pop 2"</span>), 
               <span class="dt">label.size =</span> <span class="fl">0.75</span>)</code></pre></div>
<p><img src="chavi_files/figure-html/plot-consensus-6-1.png" width="672"></p>
<p>An alternative (or complementary) visualization is a heat map of pairwise IBS segment lengths. The values on the heat map scale are segment lengths, so red colors indicate long pairwise segments and blue colors indicate shorter segments.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/ibs_heatmap.html">ibs_heatmap</a></span>(cons.out<span class="op">$</span>consensus)</code></pre></div>
<p><img src="chavi_files/figure-html/plot-heatmap-1.png" width="480"></p>
<p>This again has several options, including group organization and demarcation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/ibs_heatmap.html">ibs_heatmap</a></span>(cons.out<span class="op">$</span>consensus, <span class="dt">groups =</span> <span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">2</span>), 
            <span class="dt">group.lines =</span> T)</code></pre></div>
<p><img src="chavi_files/figure-html/plot-heatmap-2-1.png" width="480"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#step-0-install-chavi">Step 0: Install chavi</a></li>
      <li><a href="#step-1-load-and-prepare-data">Step 1: Load and prepare data</a></li>
      <li><a href="#step-2-identify-consensus-risk-haplotype">Step 2: Identify consensus risk haplotype</a></li>
      <li><a href="#step-3-visualize-consensus-risk-haplotype">Step 3: Visualize consensus risk haplotype</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Anna Plantinga.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
